jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java

```diff
@@ -0,0 +1,303 @@
+
+                if (_terminal != null && _buffer != null)
+                {
+                    _buffer.release();
+                    _buffer = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    public Content.Chunk read()
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            checkOpenLocked();
+            if (_terminal != null)
+                return _terminal;
+
+            if (_buffer == null)
+            {
+                _buffer = _byteBufferPool.acquire();
+            }
+            else if (_buffer.isRetained())
+            {
+                _buffer.release();
+                _buffer = _byteBufferPool.acquire();
+            }
+
+            try
+            {
+                ByteBuffer byteBuffer = _buffer.getByteBuffer();
+                BufferUtil.clearToFill(byteBuffer);
+                if (_length >= 0)
+                    byteBuffer.limit((int)Math.min(_buffer.capacity(), _length - _totalRead));
+                int read = _byteChannel.read(byteBuffer);
+                BufferUtil.flipToFlush(byteBuffer, 0);
+                if (read == 0)
+                    return null;
+                if (read > 0)
+                {
+                    _totalRead += read;
+                    if (_length < 0 || _totalRead < _length)
+                    {
+                        _buffer.retain();
+                        return Content.Chunk.asChunk(byteBuffer, false, _buffer);
+                    }
+
+                    _terminal = Content.Chunk.EOF;
+                    IO.close(_byteChannel);
+                    Content.Chunk last = Content.Chunk.asChunk(byteBuffer, true, _buffer);
+                    _buffer = null;
+                    return last;
+                }
+                _buffer.release();
+                _buffer = null;
+                _terminal = Content.Chunk.EOF;
+                IO.close(_byteChannel);
+            }
+            catch (Throwable t)
+            {
+                _terminal = Content.Chunk.from(t, true);
+            }
+        }
+        return _terminal;
+    }
+
+    @Override
+    public void fail(Throwable failure)
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            if (_terminal == null)
+                _terminal = Content.Chunk.from(failure, true);
+            else
+                ExceptionUtil.addSuppressedIfNotAssociated(_terminal.getFailure(), failure);
+            IO.close(_byteChannel);
+            if (_buffer != null)
+            {
+                _buffer.release();
+                _buffer = null;
+            }
+        }
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public boolean rewind()
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            checkOpenLocked();
+            if (_terminal != null || _byteChannel == null || !_byteChannel.isOpen())
+                return false;
```

Comment:
i've changed checkopen to be able to remove terminal

Reply:
I think this implementation is too complicated.
It should just reset all the relevant fields, and rely on the fact that the first call to `read()` will lazily open the channel.
This will avoid code duplication like `instanceof SeekableByteChannel` and setting the position.

See the original implementation in `PathContentSource`.